File: atms/java/atms/PlanE.java
// -*- Mode: Java -*-

// ATMS-based Envisioner for planning problems
// Translated from plan-e.lisp

// Copyright (c) 1993, Kenneth D. Forbus, Northwestern University,
// and Johan de Kleer, the Xerox Corporation.
// All rights reserved.

// See the file legal.txt for a paragraph stating scope of permission
// and disclaimer of warranty.

package atms;

import java.io.PrintWriter;
import java.util.*;

/**
 * Envisioner: generates all states and transitions, and searches for plans.
 */
public class PlanE {

    /**
     * A state transition via an operator.
     */
    public static class Transition {
        public final List<Object> opInst;
        public final Env newState;

        public Transition(List<Object> opInst, Env newState) {
            this.opInst = opInst;
            this.newState = newState;
        }
    }

    /**
     * A state and its possible transitions.
     */
    public static class StateEntry {
        public final Env state;
        public final List<Transition> transitions;

        public StateEntry(Env state, List<Transition> transitions) {
            this.state = state;
            this.transitions = transitions;
        }
    }

    /**
     * Generate all states and transitions for the planning problem.
     */
    @SuppressWarnings("unchecked")
    public static List<StateEntry> envision(Planner planner) {
        List<List<TmsNode>> choiceSets = new ArrayList<>();
        for (List<List<Object>> bs : planner.getBasisSet()) {
            List<TmsNode> nodes = new ArrayList<>();
            for (List<Object> choice : bs) {
                nodes.add(AtreData.getTmsNode(choice, planner.getAtre()));
            }
            choiceSets.add(nodes);
        }

        List<Env> states = planner.getAtre().getAtms()
                .interpretations(choiceSets, null);
        planner.getPlist().put("STATES", states);

        List<StateEntry> transTable = applyAllOperators(states, planner);
        planner.getPlist().put("TRANSITIONS", transTable);
        return transTable;
    }

    private static List<StateEntry> applyAllOperators(List<Env> states,
                                                        Planner planner) {
        List<StateEntry> result = new ArrayList<>();
        for (Env state : states) {
            List<Transition> transitions = new ArrayList<>();
            for (List<Object> opInst : planner.findApplicableOperators(state)) {
                Env newState = planner.applyOperator(state, opInst);
                if (newState != null) {
                    transitions.add(new Transition(opInst, newState));
                }
            }
            result.add(new StateEntry(state, transitions));
        }
        return result;
    }

    /**
     * Display the envisionment.
     */
    @SuppressWarnings("unchecked")
    public static void showEnvisionment(Planner planner, PrintWriter stream) {
        Object statesVal = planner.getPlist().get("STATES");
        List<Env> states = statesVal instanceof List ? (List<Env>) statesVal : null;

        if (states == null || states.isEmpty()) {
            stream.println("\nThe state space is empty.");
            stream.flush();
            return;
        }

        stream.printf("%n %d states have been generated:", states.size());
        for (Env state : states) {
            stream.printf("%n%s", state);
        }

        stream.printf("%nTransition Table:");
        Object transVal = planner.getPlist().get("TRANSITIONS");
        List<StateEntry> transTable = transVal instanceof List ?
                (List<StateEntry>) transVal : null;

        if (transTable == null || transTable.isEmpty()) {
            stream.print(" empty.");
        } else {
            for (StateEntry entry : transTable) {
                stream.printf("%n  %s: ", entry.state);
                for (Transition t : entry.transitions) {
                    stream.printf("%n   %s -> %s", t.opInst, t.newState);
                }
            }
        }
        stream.flush();
    }

    /**
     * Find a plan by searching the envisionment.
     */
    @SuppressWarnings("unchecked")
    public static List<Object> findPlan(List<Object> start, List<Object> goals,
                                          Planner planner) {
        List<Env> goalStates = planner.fetchStates(goals);
        List<Env> startStates = planner.fetchStates(start);

        if (planner.isDebugging()) {
            System.out.printf("%nInitial states are %s.", startStates);
            System.out.printf("%nGoal states are %s.", goalStates);
        }

        Object transVal = planner.getPlist().get("TRANSITIONS");
        List<StateEntry> transTable = transVal instanceof List ?
                (List<StateEntry>) transVal : new ArrayList<>();

        // Initialize queue
        LinkedList<List<Object>> queue = new LinkedList<>();
        for (Env state : startStates) {
            List<Object> path = new ArrayList<>();
            path.add(state);
            queue.add(path);
        }

        while (!queue.isEmpty()) {
            List<Object> current = queue.removeFirst();
            Env currentState = (Env) current.get(0);

            // Check if we've reached a goal state
            boolean isGoal = goalStates.contains(currentState);
            if (isGoal) {
                planner.getPlist().put("PLAN", current);
                return current;
            }

            // Find transitions for the current state
            for (StateEntry entry : transTable) {
                if (entry.state != currentState) continue;
                for (Transition trans : entry.transitions) {
                    // Avoid loops
                    boolean visited = false;
                    for (int i = 2; i < current.size(); i++) {
                        if (current.get(i) instanceof Env &&
                            current.get(i) == trans.newState) {
                            visited = true;
                            break;
                        }
                    }

                    if (!visited) {
                        if (planner.isDebugging()) {
                            System.out.printf("%n Can reach %s via %s from %s.",
                                    trans.newState, trans.opInst, currentState);
                        }
                        List<Object> newPath = new ArrayList<>();
                        newPath.add(trans.newState);
                        newPath.add(trans.opInst);
                        newPath.addAll(current);
                        queue.addLast(newPath);
                    }
                }
            }
        }

        planner.getPlist().put("PLAN", null);
        return null;
    }
}

File: atms/go/plan_e.go
// -*- Mode: Go -*-

// ATMS-based Envisioner for planning problems
// Translated from plan-e.lisp

// Copyright (c) 1993, Kenneth D. Forbus, Northwestern University,
// and Johan de Kleer, the Xerox Corporation.
// All rights reserved.

// See the file legal.txt for a paragraph stating scope of permission
// and disclaimer of warranty.

package atms

import (
	"fmt"
	"io"
	"os"
)

// Transition represents a state transition via an operator.
type Transition struct {
	OpInst   []interface{}
	NewState *Env
}

// StateEntry represents a state and its possible transitions.
type StateEntry struct {
	State       *Env
	Transitions []Transition
}

// Envision generates all states and transitions for a planning problem.
func Envision(plnpr *Plnpr) []StateEntry {
	choiceSets := make([][]*TmsNode, 0)
	for _, bs := range plnpr.BasisSet {
		var nodes []*TmsNode
		for _, choice := range bs {
			nodes = append(nodes, GetTmsNode(choice, plnpr.Atre))
		}
		choiceSets = append(choiceSets, nodes)
	}
	states := Interpretations(plnpr.Atre.ATMS, choiceSets, nil)
	plnpr.Plist["STATES"] = states

	transTable := applyAllOperators(states, plnpr)
	plnpr.Plist["TRANSITIONS"] = transTable
	return transTable
}

func applyAllOperators(states []*Env, plnpr *Plnpr) []StateEntry {
	atre := plnpr.Atre
	var result []StateEntry
	for _, state := range states {
		entry := StateEntry{State: state}
		for _, opInst := range FindApplicableOperators(state, plnpr) {
			newState := ApplyOperator(state, opInst, plnpr, atre)
			if newState != nil {
				entry.Transitions = append(entry.Transitions, Transition{
					OpInst:   opInst,
					NewState: newState,
				})
			}
		}
		result = append(result, entry)
	}
	return result
}

// ShowEnvisionment displays the generated states and transition table.
func ShowEnvisionment(plnpr *Plnpr, w io.Writer) {
	if w == nil {
		w = os.Stdout
	}
	statesVal := plnpr.Plist["STATES"]
	states, ok := statesVal.([]*Env)
	if !ok || len(states) == 0 {
		fmt.Fprintf(w, "\nThe state space is empty.")
		return
	}

	fmt.Fprintf(w, "\n %d states have been generated:", len(states))
	for _, state := range states {
		PrintEnv(state, w)
	}

	fmt.Fprintf(w, "\nTransition Table:")
	transVal := plnpr.Plist["TRANSITIONS"]
	transTable, ok := transVal.([]StateEntry)
	if !ok || len(transTable) == 0 {
		fmt.Fprintf(w, " empty.")
		return
	}

	for _, entry := range transTable {
		fmt.Fprintf(w, "\n  %v: ", entry.State)
		for _, t := range entry.Transitions {
			fmt.Fprintf(w, "\n   %v -> %v", t.OpInst, t.NewState)
		}
	}
}

// FindPlan searches the envisionment for a plan from start to goals.
func FindPlan(start []interface{}, goals []interface{}, plnpr *Plnpr) []interface{} {
	goalStates := FetchStates(goals, plnpr)
	startStates := FetchStates(start, plnpr)

	DebugPlnpr(plnpr, os.Stdout, "\nInitial states are %v.", startStates)
	DebugPlnpr(plnpr, os.Stdout, "\nGoal states are %v.", goalStates)

	// Get transitions from plist
	transVal := plnpr.Plist["TRANSITIONS"]
	transTable, _ := transVal.([]StateEntry)

	// Initialize queue with start states
	queue := make([][]interface{}, 0)
	for _, state := range startStates {
		queue = append(queue, []interface{}{state})
	}

	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]

		currentState, ok := current[0].(*Env)
		if !ok {
			continue
		}

		// Check if current state is a goal state
		isGoal := false
		for _, gs := range goalStates {
			if gs == currentState {
				isGoal = true
				break
			}
		}
		if isGoal {
			plnpr.Plist["PLAN"] = current
			return current
		}

		// Find transitions for current state
		for _, entry := range transTable {
			if entry.State != currentState {
				continue
			}
			for _, trans := range entry.Transitions {
				// Avoid loops
				alreadyVisited := false
				for i := 2; i < len(current); i++ {
					if env, ok := current[i].(*Env); ok && env == trans.NewState {
						alreadyVisited = true
						break
					}
				}
				if !alreadyVisited {
					DebugPlnpr(plnpr, os.Stdout,
						"\n Can reach %v via %v from %v.",
						trans.NewState, trans.OpInst, currentState)
					newPath := make([]interface{}, 0, len(current)+2)
					newPath = append(newPath, trans.NewState)
					newPath = append(newPath, trans.OpInst)
					newPath = append(newPath, current...)
					queue = append(queue, newPath)
				}
			}
		}
	}

	plnpr.Plist["PLAN"] = nil
	return nil
}
